### DISCUSSION TITLE
Trying to figure out how premissions work in Oqtane

### DISCUSSION THREAD

**Discussion by vnetonline** [oqtane/oqtane.framework#3256] [Q&A]
@sbwalker 

**I have given the registered users the ability to access the User Management Page with `View Module` and `Edit Module` Permission**

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/cd52b9cc-52e6-447f-a186-5cabfe48cc16)

**But when I log in as a registered user and click on the edit I can't see the Manage Menu being enabled.**

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/c3b14d1d-1a06-49eb-926b-96fa5b703245)

**What am I missing?**

**I really don't understand what is the difference between using a Policy and a Role to Authorise a user**

**Comment by sbwalker** [comment]
@vnetonline your screen indicates that you gave Registered Users the ability to access the User Management MODULE - not the User Management PAGE. If you want to allow Registered Users to manage the PAGE then you need to use the Control Panel / Edit Page and set the Permissions. Providing Edit Page access to a role means that users in that role will be able perform all functions related to the page - ie. add child pages, edit page settings, delete the page, add modules to the page, modify the pane layout for modules, enter module content, etc... It covers a broad set of functionality and therefore needs to be managed with care. So I would never recommend giving Registered Users the right to edit a page.

You have granted rights for the MODULE instance on  the page. This means that Registered Users will be able to manage the content which is related to that specific module. However it is important to note that since Oqtane is a client/server application there are both UI permissions and API permissions to consider. You have granted Registered Users the right to Edit Module - which means they will be able to access add/edit/delete UI options within the module. However you never granted API permissions (Write User, Write UserRole) so this means that when a Registered User attempts to submit a change to the server, the API permissions will restrict the activity and they will end up with an error. For example if they Edit a specific user and change the email address and click Save, the User API on the server will check if the user has the right to Write information to the database - and in your case they do not, so the action will not be performed and an error should be display in the UI.

**Reply by vnetonline** [reply]
@sbwalker So since I gave the module access to the registered users and not the page I believe the edit button shouldn't appear and clicking on it doesn't do anything 

Am I right ??

**Comment by vnetonline** [comment]
Thanks @sbwalker 

I used the registered user to just try and work out what Oqtane is doing with permissions.

there is not UI message that displays the registered user is not allowed to edit the content, however there is an error logged in the event log 

so are you saying we should use a Policy type to protect API  and a role type to protect UI

as you know I am building a membership website to share my learnings of about Oqtane … I want to protect Courses where there will be categories for course basic, intermediate and advanced and based on the level the use wil be able to view and browse the course should I be using Policy type or role type permissions ??

**Comment by sbwalker** [comment]
@vnetonline I think you may want to read this blog: https://www.oqtane.org/blog/!/55/permission-enhancements

**Reply by vnetonline** [reply]
Yup I have read that blog and looking at the code I am a bit confused and the behaviour I am seeing is making me more confused

**Reply by sbwalker** [reply]
I do not understand what you mean by Policy type and Role type permissions?

**Reply by vnetonline** [reply]
@sbwalker   

[Authorize(Policy = $"{EntityNames.User}:{PermissionNames.Write}:{RoleNames.Admin}")]

Or like this 

[Authorize(Roles = RoleNames.Registered)]

**Reply by vnetonline** [reply]
@sbwalker 

In the case of creating a policy and specifying it in ModuleInfo

Like this 

Authorize(Policy = $"{EntityNames.Course}:{PermissionNames.Browse}:{RoleNames.YearlyMember}")]

Then Will it create precision name Browse and YearlyMember role automatically and appear in the permission grid ... if it doesn't exist ??

**Comment by sbwalker** [comment]
@vnetonline when it comes to the API, both Role and Policy authorization attributes are valid - it really depends on the specific application or use case in terms of which would be more appropriate.

Role ie. [Authorize(Roles = RoleNames.Registered)]

Role authorization attributes are simple and follow the standard .NET Controller pattern. You list one or more static role names which are allowed to access an API method. Note that it is not necessary to use the RoleNames enumeration - the Role value is simply a string so it is possible to specify "Registered Users" or "RegisteredUsers,Other Users" or whatever. The limitation of the Roles approach is that they are statically defined and embedded in your code - so they cannot be modified at run-time to affect their behavior. So if you need to provide access to another role you need to modify your code and redeploy your application. However, in many cases the roles which perform a function are very well defined in advance so the likelihood of them ever changing is low, so using the Role approach is fine.

Policy - [Authorize(Policy = $"{EntityNames.User}:{PermissionNames.Write}:{RoleNames.Admin}")]

Policy authorization attributes are more powerful... and are specific to Oqtane. They allow the security behavior to be abstracted from your code and managed dynamically. They allow for more complex security rules including the ability to grant/deny roles or specific users. This is accomplished through a combination of defining an Entity Name (ie. Job, Book, Car, etc...), a Permission Name (ie. Run, Read, Drive, etc...), and a default list of roles who have the specific permission by default. Note that similar to Role authorization attributes (described above) you do not need to use the framework enumerations - this is simply a string value (ie. "Book:Read:Administrators" is valid). Generally Policies are used in conjunction with the Permission Grid which provides a user interface to allow users to dynamically manage the roles and users who have access to specific entities and permissions. All of the permission information is stored in the Permission table in the database which has an API for interacting with the data. The blog link I shared above (https://www.oqtane.org/blog/!/55/permission-enhancements) explains that there are 2 types of permissions - they can be defined at a global entity level (within a site), or at a more granular entity ID level. Page permissions in the framework are at the entity ID level as there is a need to be able to define different access levels for each page in a site. The module permissions for User Management in the original post in this thread include global entity permissions for Write User and Write UserRole - which are required in order to provide delegated access to the User API for non-Administrator roles. 

So you really need to understand your security architecture of your custom application before you choose an technical implementation approach. The main question is do you need the ability to allow security to be managed dynamically at run-time or not?

**Comment by vnetonline** [comment]
@sbwalker thanks you so much for the detailed explanation I really appreciate your insight and helping me understand how Oqtane works.
